### ISSUE TITLE
Weight Change in Dashboard is Incorrect

### ISSUE THREAD

**Issue by tango2590** [babybuddy/babybuddy#587]
Our newborn only has two weights in so far, both in grams. Initial weight was 2490g, the second weight is 2420g. It's only a 70g change, but the Statistics box on the Dashboard shows a weight change of -490.

<img width="536" alt="Screen Shot 2023-01-16 at 03 43 07" src="https://user-images.githubusercontent.com/78666662/212647207-21438af8-db3c-4eda-b01c-7c1994a77a62.png">
<img width="1523" alt="Screen Shot 2023-01-16 at 03 42 53" src="https://user-images.githubusercontent.com/78666662/212647212-474d3909-eb11-45e3-8dd2-278a53bfbb43.png">

**Comment by cdubz** [comment]
Haha yeah that is weird... it's essentially projecting as if the trend will continue over the course of a full week. Probably this statistic just should not be shown if there is less than one week of data.

The relevant code is in `babybuddy/dashboard/templatetags/cards.py`:

https://github.com/babybuddy/babybuddy/blob/2690ab487614a4bf1a8f945ba87f47bb9a9c605c/dashboard/templatetags/cards.py#L585-L605

Thanks for reporting!

**Comment by LoganBrinsmead** [comment]
I think I would like to take a try with this issue.

**Comment by cdubz** [comment]
Yeah that's the bug -- once there is seven or more days between two entries it will just report the actual average.

**Comment by rom1dep** [comment]
> it will just report the actual average.

…and just to be clear, that's precisely what I have a problem with. For several reasons:

- Reporting projections as facts may discourage from doing regular checks/measurements, giving a false sense of confidence and control
- This can in turn hide developmental issues that this app stands to detect and avoid
- Even if reporting projections as facts was OK, using averages in this context is painfully wrong: growth is non-linear, and a correct estimate "under ideal circumstances" (well fed, well rested, no medical condition, …) would require navigating [growth charts](https://www.cdc.gov/growthcharts/), which I'm convinced is a good addition to babybuddy, but is a big can of worms on its own (I'm not aware that they exist as open data, they are country-dependent and are revised over time, etc).

IMO this (and other similar) box(es) should contain something like: "Weight change: -70g over 2 days (4 days ago)"

**Comment by cdubz** [comment]
> * This can in turn hide developmental issues that this app stands to detect and avoid
> 
> * Even if reporting projections as facts was OK, using averages in this context is painfully wrong: growth is non-linear, and a correct estimate "under ideal circumstances" (well fed, well rested, no medical condition, …) would require navigating [growth charts](https://www.cdc.gov/growthcharts/), which I'm convinced is a good addition to babybuddy, but is a big can of worms on its own (I'm not aware that they exist as open data, they are country-dependent and are revised over time, etc).

These are both good and valid points that I think don't have any workable solutions currently, unfortunately.

As for what is currently reported vs. what _should_ be reported I'd love any help you or anyone else who has time and interest to provide. The current "stats" stuff is definitely not rooted in any philosophy or research -- and I agree it should be 👍

**Comment by rom1dep** [comment]
Here's how it looks after the change proposed in the linked PR:
![image](https://user-images.githubusercontent.com/1933050/213322464-bf01354e-e870-498e-a484-d86ed097cee8.png)

**Comment by pamod-madubashana** [comment]
# PR Summary — Issue #587: Weight Change in Dashboard is Incorrect

## Analysis (no code changed)

**Reported symptom:** Newborn with two weight entries (2490g → 2420g, i.e. a 70g
loss) sees "Weight change" of **-490** in the Dashboard Statistics box.

**Code involved:** `dashboard/templatetags/cards.py` — `_weight_statistics()`
(lines ~738-758), with identical logic in `_height_statistics()`,
`_head_circumference_statistics()`, and `_bmi_statistics()`:

```python
weight_change = newest.weight - oldest.weight          # -70 in the report
weeks = (newest.date - oldest.date).days / 7           # 1 day -> 1/7 week
weight["change_weekly"] = weight_change / weeks        # -70 / (1/7) = -490
```

displayed by `dashboard/templates/cards/statistics.html` under the title
"Weight change per week".

**Likely explanation (verified against code, dates inferred):**
`-70 / (1/7) = -490` exactly, so the reporter's two entries were almost
certainly **1 day apart** and the displayed value is the arithmetically correct
*per-week rate* (daily loss × 7), not the total change. The user expected the
total change (-70). So the "bug" is a semantics/presentation problem, not a
simple arithmetic typo:
- Extrapolating a 1-day delta to a weekly rate amplifies noise and alarms
  (a 70g newborn loss projected as -490g/week).
- `Weight.weight` is a unit-less `FloatField` (percentile tables use kg), while
  the reporter talks in grams — a possible unit-entry confusion contributing to
  the surprise.
- Related latent crash in the same lines: if the two entries share the same
  `date` (plausible for a newborn weighed twice in one day),
  `(newest.date - oldest.date).days == 0` → `weeks == 0` →
  `ZeroDivisionError` (500 on the dashboard). The `if newest != oldest`
  guard compares rows by pk, not dates, so it does not protect this.

**Why not fixed unattended (ambiguous / needs product input):**
- Changing the statistic's meaning (total change vs. per-week rate), clamping
  the divisor (e.g. `max(weeks, 1)`), enforcing a minimum observation window
  (e.g. "Not enough data" under 7 days), or re-labeling the card are all
  product/design decisions affecting weight, height, head-circumference, and
  BMI cards plus existing test expectations (`test_card_statistics` expects
  `1.0` over exactly 7 days, which any clamping must preserve).
- The correct unit story (kg vs. g, label showing units) is also undecided.
- A zero-division guard alone would be safe but would **not** resolve the
  reported -490, so shipping it as "the fix for #587" would be misleading.

## Suggested approach (for a human follow-up)
1. Decide the intended statistic: keep "per week" rate or switch to total
   change since first/previous entry; update the card title accordingly and
   include units.
2. Guard short intervals: e.g. return `0.0`/"Not enough data" when
   `weeks == 0` (fixes the crash), and consider only annualizing when the span
   is ≥ 7 days (or divide by exact day count with rounding) so 1-day deltas
   don't display ×7.
3. Apply the same fix to all four sibling functions (weight, height,
   head circumference, BMI) and extend
   `dashboard/tests/tests_templatetags.py::test_card_statistics` with cases:
   two entries 1 day apart, two entries on the same date, single entry.
4. Optionally clarify weight units in the form/list (kg) to avoid g/kg mixups.

**Comment by TianHengZhuang** [comment]
read through the dashboard stats code

weekly change was projecting a full week from short spans (1 day / -70g -> ~-490). opened a PR that only reports the weekly rate when oldest/newest are >= 7 days apart, same guard on height/head circumference/BMI. #589 is still a draft from 2023 so not starting from that branch
